commands/list_keypairs.go

```diff
@@ -23,84 +27,208 @@ var (
 
 	// ListKeypairsCommand performs the "list keypairs" function
 	ListKeypairsCommand = &cobra.Command{
-		Use:     "keypairs",
-		Short:   "List key pairs for a cluster",
-		Long:    `Prints a list of key pairs for a cluster`,
-		PreRunE: checkListKeypairs,
-		Run:     listKeypairs,
+		Use:    "keypairs",
+		Short:  "List key pairs for a cluster",
+		Long:   `Prints a list of key pairs for a cluster`,
+		PreRun: listKeypairsValidationOutput,
+		Run:    listKeypairsOutput,
 	}
 )
 
+// listKeypairsArguments are the actual arguments used to call the
+// listKeypairs() function.
+type listKeypairsArguments struct {
+	apiEndpoint string
+	clusterID   string
+	token       string
+}
+
+// defaultListKeypairsArguments returns a new listKeypairsArguments struct
+// based on global variables (= command line options from cobra).
+func defaultListKeypairsArguments() listKeypairsArguments {
+	return listKeypairsArguments{
+		apiEndpoint: cmdAPIEndpoint,
+		clusterID:   cmdClusterID,
+		token:       cmdToken,
+	}
+}
+
+// listKeypairsResult is the data structure returned by the listKeypairs() function.
+type listKeypairsResult struct {
+	keypairs []gsclientgen.KeyPairModel
+}
+
 func init() {
 	ListKeypairsCommand.Flags().StringVarP(&cmdClusterID, "cluster", "c", "", "ID of the cluster to list key pairs for")
 	ListCommand.AddCommand(ListKeypairsCommand)
 }
 
-func checkListKeypairs(cmd *cobra.Command, args []string) error {
-	if config.Config.Token == "" {
-		return errors.New("You are not logged in. Use '" + config.ProgramName + " login' to log in.")
+// listKeypairsValidationOutput does our pre-checks and shows errors, in case
+// something is missing.
+func listKeypairsValidationOutput(cmd *cobra.Command, extraArgs []string) {
+	args := defaultListKeypairsArguments()
+	err := listKeypairsValidate(&args)
+	if err != nil {
+		var headline string
+		var subtext string
+
+		switch err.Error() {
+		case errNotLoggedIn:
```

Comment:
This is rather dangerous. As soon as there is a typo in the error message or it changes the error assertion will fail because we only check for the error string. We should rather use type based error assertions like we do in all of our other projects. Check some microservice and see how it is done.